### DISCUSSION TITLE
GDScript documentation comments for constructors

### DISCUSSION THREAD

**Discussion by Iakobs** [godotengine/godot-proposals#11276] [Editor]
Hi, I'm not sure if this discussion belongs to this section, but I guess it's fine.

The thing is that with [GDScript documentation comments](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_documentation_comments.html) you can document constructors, but they appear in the `Methods` section of the generated documentation, instead of in a section of their own, like native GDScript types do (e.g. the [String](https://docs.godotengine.org/en/4.3/classes/class_string.html#constructors) class).

I've searched the internet to see if this is achieved somehow but I couldn't find anything. Does this make sense as a feature proposal?

**Comment by dalexeev** [comment]
Constructors are placed in a separate section only for built-in types (`Vector2`, `Color`, etc.). I think there are some reasons for this:

- Objects are different from built-in types in that you need to use `.new()` to call a constructor. E.g. `Vector2()` and `Node.new()`.
- `_init()` is more of an initialization hook than a constructor. Class fields are initialized before `_init()` is called, and since class field initializers support non-constant expressions, you can call any method in them. Since some method can be called before `_init()`, it is not very correct to call it a constructor. The constructor is `new()`, but you cannot override it directly.
- Documentation of other programming languages ​​also does not always put constructors or initialization hooks in a separate section. For example, in PHP documentation `__construct()` is located among other methods ([example](https://www.php.net/manual/en/class.datetime.php)).

**Reply by Iakobs** [reply]
that makes sense! thanks for the clarification, I was suspecting something similar and that's why I preferred to open a discussion before proposing the feature.

Should I close the discussion now?

**Reply by dalexeev** [reply]
Yes, you can close the discussion if you are no longer interested in the proposal. But I would suggest waiting for other opinions. Maybe @AThousandShips and @Mickeon are interested as members of the documentation team?

**Reply by AThousandShips** [reply]
I'd say there isn't a way to declare constructors for custom types (in fact only built-in types have them as mentioned above) so I think it's a bit of a situation that isn't applicable 

If the engine changed to allow constructors they'd fit in a distinct section though I'd say

**Reply by Mickeon** [reply]
I don't know. I think a section for the constructor makes sense, as one of the most important parts of a class. But currently that would have to be hardcoded _specifically_ for custom classes (by fetching `_init`'s signature and description), as built-in classes do not have this luxury.

**Reply by Iakobs** [reply]
I understand all this points, and I'm pretty much inclined to just close the discussion. Just one more question, what about a system to create user defined sections inside the generate docs, like the group annotations in the export variables? Maybe it's too much for this feature.

**Reply by AThousandShips** [reply]
I'd suggest opening a new discussion or better yet a proposal for that, that is an idea I do like myself, it would be useful for the general documentation itself and I'd be interested in looking at implementing it for the class bindings side (not familiar enough with the GDScript documentation code), assuming it gets popular

I can absolutely imagine this being useful for various classes with a lot of content

It'd be especially useful for things like `@Global`

**Reply by dalexeev** [reply]
Closing per the above comments. Thanks for the contribution!
